src/GameEngine/GameEngine.cpp

```diff
@@ -0,0 +1,284 @@
+        select_map();
+        create_deck();
+        select_player();
+        Terminal::debug("Game Engine Done with start part.");
+    }
+
+    void GameEngine::start_test(int map_index, int nb_player) {
+        nb_player -= 1; // needed since 3 players = choice nb 2;
+        Terminal::debug("Game Engine Starting...");
+        Terminal::set_input(map_index);
+        select_map();
+        create_deck();
+        Terminal::set_input(nb_player);
+        select_player();
+        Terminal::debug("Game Engine Done.");
+    }
+
+    GameEngine::GameEngine() {
+        map = nullptr;
+        deck = nullptr;
+        players = new vector<Player::Player *>();
+        player_order = new vector<int>();
+        // map_registry = new Map::MapRegistry();
+    }
+
+    GameEngine::~GameEngine() {
+        delete map;
+        delete players;
+        delete deck;
+        delete player_order;
+        player_order = nullptr;
+        deck = nullptr;
+        map = nullptr;
+        players = nullptr;
+    }
+
+    Map * GameEngine::get_map() {
+        return map;
+    }
+
+    vector<Player::Player *> * GameEngine::get_players() {
+        return players;
+    }
+
+    Cards::Deck *GameEngine::get_deck() {
+        return deck;
+    }
+
+    void GameEngine::reset_test() {
+        delete map;
+        delete players;
+        delete deck;
+        delete player_order;
+        map = nullptr;
+        deck = nullptr;
+        player_order = new vector<int>();
+        players = new vector<Player::Player *>();
+    }
+
+    vector<int> *GameEngine::get_player_order() {
+        return player_order;
+    }
+
+    void GameEngine::startup_phase() {
+        assign_player_order_randomly();
+        assign_country_to_player();
+        assign_armies_into_country();
+    }
+
+    void GameEngine::assign_player_order_randomly() {
+        Terminal::debug("Determining player order");
+
+        // Seeding random function for unpredictable results
+        srand(time(0));
+
+        for (auto player : *players){
+            bool unique = true;
+            int order = 0;
+            do {
+                order = rand() % players->size();
+                unique = true;
+                for (auto i : *player_order) {
+                    if (order == i) {
+                        unique = false;
+                        break;
+                    }
+                }
+            } while(!unique);
+            player_order->emplace_back(order);
+        }
+
+        Terminal::debug("The order is");
+        for (auto order : * player_order)
+            Terminal::debug("Player " + to_string(order+1));
+    }
+
+    void GameEngine::assign_country_to_player() {
+        // Loop countries with player index increasing
+        vector<Country *> countries = map->get_countries();
+
+        shuffle( countries.begin(), countries.end(), default_random_engine(time(0)));
+
+        int owner_index = 0;
+        for(auto & countrie : countries){
```

Comment:
countrie sounds cute, but it's spelled country 😄

Reply:
Lol IDE auto-completed I guess XD thx